### PR TITLE
Free WT_REFs deleted by truncate

### PR DESCRIPTION
**Pull request by michaelcahill**
We were doing this when a page spontaneously became empty, but not if the "fast truncate" code kicked in.

refs SERVER-16921

### CODE REVIEW

**Comment by wtbuild** [comment]
Test PASSed.

**Comment by michaelcahill** [comment]
@keithbostic, the reason we didn't do this originally was that I was concerned that we might see a `WT_REF` that was deleted by truncate, but `page_del->txnid` may not be globally visible.

I *think* we only get here from eviction, so `isolation == WT_ISO_EVICTION` and the visibility check in `__wt_delete_page_skip` is sufficient.  If that is not the case, so we could be doing a split without coming through eviction, then we need to pass through a flag to make that an explicit `__wt_txn_visible_all`.

**Comment by michaelcahill** [comment]
Also, fwiw, this test makes the single-threaded test case in 16921 run 15-20% faster over the first 10M inserts.  I think that gap will only get larger the longer the workload keeps running.

**Comment by keithbostic** [comment]
We can also get here if the file is being closed, `__wt_evict_file(WT_SYNC_CLOSE)` calls `__wt_evict`, which can end up in this code. I believe that's safe, but wanted to call it out.

File: src/btree/bt_split.c [hunk 1]
```diff
@@ -986,6 +985,18 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
...
+				 * immediately: it is only read when the ref
+				 * state is WT_REF_DELETED.  The size of the
+				 * structures wasn't added to the parent: don't
+				 * decrement.
+				 */
+				if (next_ref->page_del != NULL) {
+					__wt_free(session,
+					    next_ref->page_del->update_list);
+					__wt_free(session, next_ref->page_del);
+				}
```
Comment (keithbostic) [inline src/btree/bt_split.c]: @michaelcahill, I think this code should appear at line 893 where we flip the WT_REF state, we never reset the state on failure, so we should clean up the memory at the same time. Something like this:
```
diff --git a/src/btree/bt_split.c b/src/btree/bt_split.c
index 018136d..e64f2d5 100644
--- a/src/btree/bt_split.c
+++ b/src/btree/bt_split.c
@@ -893,6 +893,19 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
                    WT_ATOMIC_CAS4(next_ref->state,
                    WT_REF_DELETED, WT_REF_SPLIT))
                        deleted_entries++;
+
+                       /*
+                        * The page_del structure is freed immediately: it is
+                        * only read when the ref state is WT_REF_DELETED and
+                        * we do not reset that value on failure. The size of
+                        * the structures wasn't added to the parent: don't
+                        * decrement.
+                        */
+                       if (next_ref->page_del != NULL) {
+                               __wt_free(session,
+                                   next_ref->page_del->update_list);
+                               __wt_free(session, next_ref->page_del);
+                       }
        }
 
        /*
@@ -985,18 +998,6 @@ __split_parent(WT_SESSION_IMPL *session, WT_REF *ref, WT_REF **ref_new,
                                            session, 0, ikey, size));
                                        WT_MEMSIZE_ADD(parent_decr, size);
                                }
-                               /*
-                                * The page_del structure can be freed
-                                * immediately: it is only read when the ref
-                                * state is WT_REF_DELETED.  The size of the
-                                * structures wasn't added to the parent: don't
-                                * decrement.
-                                */
-                               if (next_ref->page_del != NULL) {
-                                       __wt_free(session,
-                                           next_ref->page_del->update_list);
-                                       __wt_free(session, next_ref->page_del);
-                               }
                        }
 
                        WT_TRET(__split_safe_free(
```

**Comment by keithbostic** [comment]
@michaelcahill, this looks good to me with one suggested code change, let's merge it.

File: src/btree/bt_split.c (hunk 1 above)
Comment (michaelcahill) [inline src/btree/bt_split.c]: We do restore the state, at line 1067:

```c
err:    if (!complete)
                for (i = 0; i < parent_entries; ++i) {
                        next_ref = pindex->index[i];
                        if (next_ref->state == WT_REF_SPLIT)
                                next_ref->state = WT_REF_DELETED;
                }
```

Given that, I think we should free `page_del` at the same time we schedule other memory on the page to be freed (that is, after the split is completed and we're never going to restore the ref states).

Comment (keithbostic) [inline src/btree/bt_split.c]: Got it -- sorry, I missed that.

Still... there's no reason to keep the memory around, is there?

Either way, I'm good to go.

**Pull request merged**